spec/lib/diaspora_federation/discovery/discovery_spec.rb

```diff
@@ -88,66 +91,148 @@ module DiasporaFederation
           callback_person = person
         end
 
-        expect(Discovery::Discovery.new(account).fetch_and_save).to be(callback_person)
+        expect(subject.fetch_and_save).to be(callback_person)
       end
 
       it "falls back to http if https fails with 404" do
-        stub_request(:get, "https://localhost:3000/.well-known/host-meta")
+        stub_request(:get, "https://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
           .to_return(status: 404)
-        stub_request(:get, "http://localhost:3000/.well-known/host-meta")
-          .to_return(status: 200, body: host_meta_xrd)
-        stub_request(:get, "http://localhost:3000/.well-known/webfinger.xml?resource=acct:#{account}")
-          .to_return(status: 200, body: webfinger_xrd)
+        stub_request(:get, "http://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
+          .to_return(status: 200, body: webfinger_jrd)
         stub_request(:get, "http://localhost:3000/hcard/users/#{alice.guid}")
           .to_return(status: 200, body: hcard_html)
 
         expect_callback(:save_person_after_webfinger, kind_of(Entities::Person))
-        person = Discovery::Discovery.new(account).fetch_and_save
+        person = subject.fetch_and_save
 
         expect(person.guid).to eq(alice.guid)
         expect(person.diaspora_id).to eq(account)
       end
 
       it "falls back to http if https fails with ssl error" do
```

Comment:
Similar to the above, falling back to https we actually allow MitM attack. If :imp: wants to serve something instead of the original resource he just replies to the https request with self-signed certificate and then just passes desired data via http.

Reply:
Same as above, it was always that way. But it's only for discovery, once a contact is known, there is no possibility to change the key or force it back to http when it was once https.

If you think we should change something here it's probably the best when you open a discussion on discourse. But I think it's out of scope here (because as said, it's currently that way, and we would at least need a major release to remove http compatibility)